File: mobile/modules/bluetooth-sdk/ios/Source/utils/BleJsonCompact.swift
import Foundation

/// BLE Wire Protocol v2 JSON compaction: short keys, enum ints, omitted defaults, config diff.
enum BleJsonCompact {
    static let keyResolvedConfigHash = "rch"

    private static let longToShort: [String: String] = [
        "type": "t",
        "requestId": "r",
        "timestamp": "ts",
        "status": "s",
        "kind": "k",
        "source": "src",
        "requestedCaptureConfig": "rcc",
        "meteredPreview": "mp",
        "exposureTimeNs": "etn",
        "captureMetadata": "cm",
        "aeStateName": "aes",
        "transferMethod": "tm",
        "manual": "m",
        "reconnecting": "rc",
        "stats": "st",
        "width": "w",
        "height": "h",
        "bitrate": "br",
        "fps": "f",
        "droppedFrames": "df",
        "duration": "du",
        "temperatureC": "tc",
    ]

    private static let shortToLong: [String: String] = {
        var map: [String: String] = [:]
        for (longKey, shortKey) in longToShort {
            map[shortKey] = longKey
        }
        return map
    }()

    private static let photoStatusToInt: [String: Int] = [
        "accepted": 0, "queued": 1, "configuring": 2, "capturing": 3,
        "captured": 4, "uploading": 5, "uploaded": 6, "failed": 7,
    ]
    private static let photoStatusFromInt: [Int: String] = {
        var map: [Int: String] = [:]
        for (key, value) in photoStatusToInt { map[value] = key }
        return map
    }()

    private static let kindToInt = ["lifecycle": 0, "snapshot": 1]
    private static let kindFromInt: [Int: String] = [0: "lifecycle", 1: "snapshot"]

    private static let sourceToInt = ["sdk": 0, "button": 1]
    private static let sourceFromInt: [Int: String] = [0: "sdk", 1: "button"]

    private static let aeStateToInt: [String: Int] = [
        "CONVERGED": 0, "SEARCHING": 1, "INACTIVE": 2,
        "LOCKED": 3, "FLASH_REQUIRED": 4, "PRECAPTURE": 5,
    ]
    private static let aeStateFromInt: [Int: String] = {
        var map: [Int: String] = [:]
        for (key, value) in aeStateToInt { map[value] = key }
        return map
    }()

    private static var sessionConnectEpochMs: Int64 = 0
    private static var resolvedConfigSent = false
    private static var currentSessionResolvedConfigHash: String?
    private static var resolvedConfigByHash: [String: [String: Any]] = [:]

    private static let highRoiMessageTypes: Set<String> = [
        "photo_status",
        "stream_status",
        "wifi_scan_result",
        "start_stream",
        "photo_response",
    ]

    private static let chunkMessageTypes: Set<String> = ["ck", "chunked_msg"]

    static func shouldCompactOutbound(_ messageType: String) -> Bool {
        highRoiMessageTypes.contains(messageType)
    }

    static func supportsCompactInbound(_ messageType: String) -> Bool {
        !messageType.isEmpty
            && (highRoiMessageTypes.contains(messageType) || chunkMessageTypes.contains(messageType))
    }

    static func isCompactWireForm(_ json: [String: Any]) -> Bool {
        json["t"] != nil && json["type"] == nil
    }

    static func extractMessageType(_ json: [String: Any]) -> String {
        if let type = stringValue(json["type"]) {
            return type
        }
        if let compactType = stringValue(json["t"]) {
            return compactType
        }
        return ""
    }

    static func resetSession() {
        sessionConnectEpochMs = 0
        resolvedConfigSent = false
        currentSessionResolvedConfigHash = nil
        resolvedConfigByHash.removeAll()
    }

    static func markSessionConnected(epochMs: Int64) {
        resetSession()
        sessionConnectEpochMs = epochMs
    }

    static func isCameraCommandJson(_ jsonData: String) -> Bool {
        jsonData.contains("cs_pho")
            || jsonData.contains("cs_cpho")
            || jsonData.contains("cs_vid")
            || jsonData.contains("\"type\":\"take_photo\"")
    }

    static func encode(_ json: [String: Any]) -> [String: Any] {
        if isCameraCommandJson(jsonString(json) ?? "") {
            return json
        }
        let messageType = extractMessageType(json)
        if !shouldCompactOutbound(messageType) {
            return json
        }
        return compactObject(json, topLevel: true)
    }

    static func encode(jsonString: String) -> [String: Any]? {
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return nil
        }
        return encode(json)
    }

    static func decode(_ json: [String: Any]) -> [String: Any] {
        if isCameraCommandJson(jsonString(json) ?? "") {
            return json
        }
        return expandObject(json, topLevel: true)
    }

    static func decodeIfSupported(_ json: [String: Any]) -> [String: Any]? {
        if isCameraCommandJson(jsonString(json) ?? "") {
            return json
        }
        if !isCompactWireForm(json) {
            return json
        }
        let compactType = stringValue(json["t"]) ?? ""
        if !supportsCompactInbound(compactType) {
            return nil
        }
        return expandObject(json, topLevel: true)
    }

    static func decode(jsonString: String) -> [String: Any]? {
        guard let data = jsonString.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return nil
        }
        return decode(json)
    }

    private static func compactObject(_ input: [String: Any], topLevel: Bool) -> [String: Any] {
        var output: [String: Any] = [:]
        let messageType = stringValue(input["type"]) ?? stringValue(input["t"]) ?? ""
        var skipResolvedConfig = false

        if topLevel, shouldDiffResolvedConfig(messageType),
           let resolvedConfig = input["resolvedConfig"] as? [String: Any]
        {
            let hash = hashConfig(resolvedConfig)
            resolvedConfigByHash[hash] = resolvedConfig
            if resolvedConfigSent, hash == currentSessionResolvedConfigHash {
                output[keyResolvedConfigHash] = hash
                skipResolvedConfig = true
            } else {
                currentSessionResolvedConfigHash = hash
                resolvedConfigSent = true
            }
        }

        for (key, value) in input {
            if skipResolvedConfig, key == "resolvedConfig" { continue }
            let shortKey = longToShort[key] ?? key
            guard let compacted = compactValue(longKey: key, value: value, messageType: messageType),
                  !shouldOmit(compacted)
            else {
                continue
            }
            output[shortKey] = compacted
        }

        if topLevel {
            compactTimestamp(&output)
        }
        return output
    }

    private static func shouldDiffResolvedConfig(_ messageType: String) -> Bool {
        messageType == "photo_status" || messageType == "stream_status"
    }

    private static func compactValue(longKey: String, value: Any, messageType: String) -> Any? {
        if value is NSNull { return nil }
        if let nested = value as? [String: Any] {
            return compactObject(nested, topLevel: false)
        }
        if let array = value as? [Any] {
            return array.map { item -> Any in
                if let nested = item as? [String: Any] {
                    return compactObject(nested, topLevel: false)
                }
                return item
            }
        }
        if let boolValue = value as? Bool { return boolValue }
        if let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID() {
            return number
        }
        if let str = value as? String {
            if longKey == "status", messageType == "photo_status", let code = photoStatusToInt[str] {
                return code
            }
            if longKey == "kind", let code = kindToInt[str] { return code }
            if longKey == "source", let code = sourceToInt[str] { return code }
            if longKey == "aeStateName", let code = aeStateToInt[str] { return code }
            return str
        }
        return value
    }

    private static func shouldOmit(_ value: Any) -> Bool {
        if value is NSNull { return true }
        if let boolValue = value as? Bool, !boolValue { return true }
        if let dict = value as? [String: Any], dict.isEmpty { return true }
        if let array = value as? [Any], array.isEmpty { return true }
        return false
    }

    private static func compactTimestamp(_ output: inout [String: Any]) {
        let tsKey: String?
        if output["timestamp"] != nil {
            tsKey = "timestamp"
        } else if output["ts"] != nil {
            tsKey = "ts"
        } else {
            tsKey = nil
        }
        guard let key = tsKey else { return }
        let absolute = int64Value(output[key]) ?? 0
        guard absolute > 0 else { return }
        if sessionConnectEpochMs <= 0 {
            sessionConnectEpochMs = absolute
        }
        output["ts"] = absolute - sessionConnectEpochMs
        if key != "ts" {
            output.removeValue(forKey: key)
        }
    }

    private static func expandObject(_ input: [String: Any], topLevel: Bool) -> [String: Any] {
        var output: [String: Any] = [:]
        let messageType = stringValue(input["type"]) ?? stringValue(input["t"]) ?? ""

        for (key, value) in input {
            if topLevel, key == "ts" {
                output["ts"] = value
                continue
            }
            let longKey = shortToLong[key] ?? key
            guard let expanded = expandValue(longKey: longKey, value: value, messageType: messageType) else {
                continue
            }
            output[longKey] = expanded
        }

        if topLevel {
            expandTimestamp(&output)
            expandResolvedConfigHash(&output)
            cacheResolvedConfigIfPresent(&output)
        }
        return output
    }

    private static func cacheResolvedConfigIfPresent(_ output: inout [String: Any]) {
        guard let resolvedConfig = output["resolvedConfig"] as? [String: Any] else { return }
        let hash = hashConfig(resolvedConfig)
        resolvedConfigByHash[hash] = resolvedConfig
        currentSessionResolvedConfigHash = hash
        resolvedConfigSent = true
    }

    private static func expandResolvedConfigHash(_ output: inout [String: Any]) {
        guard output["resolvedConfig"] == nil,
              let hash = output[keyResolvedConfigHash] as? String,
              let cached = resolvedConfigByHash[hash]
        else {
            return
        }
        output["resolvedConfig"] = cached
        output.removeValue(forKey: keyResolvedConfigHash)
    }

    private static func expandValue(longKey: String, value: Any, messageType: String) -> Any? {
        if value is NSNull { return nil }
        if let nested = value as? [String: Any] {
            return expandObject(nested, topLevel: false)
        }
        if let array = value as? [Any] {
            return array.map { item -> Any in
                if let nested = item as? [String: Any] {
                    return expandObject(nested, topLevel: false)
                }
                return item
            }
        }
        if longKey == "status", messageType == "photo_status", let code = value as? Int {
            return photoStatusFromInt[code] ?? value
        }
        if longKey == "kind", let code = value as? Int {
            return kindFromInt[code] ?? value
        }
        if longKey == "source", let code = value as? Int {
            return sourceFromInt[code] ?? value
        }
        if longKey == "aeStateName", let code = value as? Int {
            return aeStateFromInt[code] ?? value
        }
        return value
    }

    private static func expandTimestamp(_ output: inout [String: Any]) {
        guard output["timestamp"] == nil, let delta = int64Value(output["ts"]) else { return }
        let absolute = sessionConnectEpochMs > 0 ? sessionConnectEpochMs + delta : delta
        output["timestamp"] = absolute
        output.removeValue(forKey: "ts")
        if sessionConnectEpochMs <= 0 {
            sessionConnectEpochMs = absolute - delta
        }
    }

    static func hashConfig(_ config: [String: Any]) -> String {
        fnv1a32Hex(canonicalJsonString(config))
    }

    private static func canonicalJsonString(_ object: [String: Any]) -> String {
        let sortedKeys = object.keys.sorted()
        var parts: [String] = []
        for key in sortedKeys {
            guard let value = object[key] else { continue }
            let encodedValue: String
            if let nested = value as? [String: Any] {
                encodedValue = canonicalJsonString(nested)
            } else if let str = value as? String {
                encodedValue = "\"\(str)\""
            } else if let boolValue = value as? Bool {
                encodedValue = boolValue ? "true" : "false"
            } else if let number = value as? NSNumber {
                encodedValue = "\(number)"
            } else {
                encodedValue = "null"
            }
            parts.append("\"\(key)\":\(encodedValue)")
        }
        return "{\(parts.joined(separator: ","))}"
    }

    private static func fnv1a32Hex(_ value: String) -> String {
        var hash: UInt32 = 0x811c9dc5
        for byte in value.utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 0x01000193
        }
        return String(format: "%08x", hash)
    }

    private static func stringValue(_ value: Any?) -> String? {
        if let str = value as? String { return str }
        if let number = value as? NSNumber { return number.stringValue }
        return nil
    }

    private static func int64Value(_ value: Any?) -> Int64? {
        if let number = value as? NSNumber { return number.int64Value }
        if let str = value as? String, let parsed = Int64(str) { return parsed }
        return nil
    }

    private static func jsonString(_ json: [String: Any]) -> String? {
        guard let data = try? JSONSerialization.data(withJSONObject: json) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}

File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/utils/BleJsonCompact.java
package com.mentra.bluetoothsdk.utils;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/** BLE Wire Protocol v2 JSON compaction: short keys, enum ints, omitted defaults, config diff. */
public final class BleJsonCompact {

    private static final Map<String, String> LONG_TO_SHORT = new HashMap<>();
    private static final Map<String, String> SHORT_TO_LONG = new HashMap<>();

    private static final Map<String, Integer> PHOTO_STATUS_TO_INT = new HashMap<>();
    private static final Map<Integer, String> PHOTO_STATUS_FROM_INT = new HashMap<>();
    private static final Map<String, Integer> KIND_TO_INT = new HashMap<>();
    private static final Map<Integer, String> KIND_FROM_INT = new HashMap<>();
    private static final Map<String, Integer> SOURCE_TO_INT = new HashMap<>();
    private static final Map<Integer, String> SOURCE_FROM_INT = new HashMap<>();
    private static final Map<String, Integer> AE_STATE_TO_INT = new HashMap<>();
    private static final Map<Integer, String> AE_STATE_FROM_INT = new HashMap<>();

    public static final String KEY_RESOLVED_CONFIG_HASH = "rch";

    /** ROI 8–10 message types: compact outbound and accept compact inbound. */
    private static final Set<String> HIGH_ROI_MESSAGE_TYPES =
            Collections.unmodifiableSet(
                    new HashSet<>(
                            Arrays.asList(
                                    "photo_status",
                                    "stream_status",
                                    "wifi_scan_result",
                                    "start_stream",
                                    "photo_response")));

    /** Structural chunk envelope (always compact on wire). */
    private static final Set<String> CHUNK_MESSAGE_TYPES =
            Collections.unmodifiableSet(new HashSet<>(Arrays.asList("ck", "chunked_msg")));

    private static long sessionConnectEpochMs;
    private static boolean resolvedConfigSent;
    private static String currentSessionResolvedConfigHash;
    private static final Map<String, JSONObject> resolvedConfigByHash = new HashMap<>();

    static {
        putKey("type", "t");
        putKey("requestId", "r");
        putKey("timestamp", "ts");
        putKey("status", "s");
        putKey("kind", "k");
        putKey("source", "src");
        putKey("requestedCaptureConfig", "rcc");
        putKey("meteredPreview", "mp");
        putKey("exposureTimeNs", "etn");
        putKey("captureMetadata", "cm");
        putKey("aeStateName", "aes");
        putKey("transferMethod", "tm");
        putKey("manual", "m");
        putKey("reconnecting", "rc");
        putKey("stats", "st");
        putKey("width", "w");
        putKey("height", "h");
        putKey("bitrate", "br");
        putKey("fps", "f");
        putKey("droppedFrames", "df");
        putKey("duration", "du");
        putKey("temperatureC", "tc");

        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "accepted", 0);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "queued", 1);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "configuring", 2);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "capturing", 3);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "captured", 4);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "uploading", 5);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "uploaded", 6);
        putEnum(PHOTO_STATUS_TO_INT, PHOTO_STATUS_FROM_INT, "failed", 7);

        putEnum(KIND_TO_INT, KIND_FROM_INT, "lifecycle", 0);
        putEnum(KIND_TO_INT, KIND_FROM_INT, "snapshot", 1);

        putEnum(SOURCE_TO_INT, SOURCE_FROM_INT, "sdk", 0);
        putEnum(SOURCE_TO_INT, SOURCE_FROM_INT, "button", 1);

        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "CONVERGED", 0);
        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "SEARCHING", 1);
        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "INACTIVE", 2);
        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "LOCKED", 3);
        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "FLASH_REQUIRED", 4);
        putEnum(AE_STATE_TO_INT, AE_STATE_FROM_INT, "PRECAPTURE", 5);
    }

    private BleJsonCompact() {}

    public static void resetSession() {
        sessionConnectEpochMs = 0L;
        resolvedConfigSent = false;
        currentSessionResolvedConfigHash = null;
        resolvedConfigByHash.clear();
    }

    public static void markSessionConnected(long epochMs) {
        resetSession();
        sessionConnectEpochMs = epochMs;
    }

    public static boolean isCameraCommandJson(String jsonData) {
        if (jsonData == null || jsonData.isEmpty()) {
            return false;
        }
        return jsonData.contains("cs_pho")
                || jsonData.contains("cs_cpho")
                || jsonData.contains("cs_vid")
                || jsonData.contains("\"type\":\"take_photo\"");
    }

    public static boolean shouldCompactOutbound(String messageType) {
        return messageType != null && HIGH_ROI_MESSAGE_TYPES.contains(messageType);
    }

    public static boolean supportsCompactInbound(String messageType) {
        if (messageType == null || messageType.isEmpty()) {
            return false;
        }
        return HIGH_ROI_MESSAGE_TYPES.contains(messageType)
                || CHUNK_MESSAGE_TYPES.contains(messageType);
    }

    /** Compact wire form uses short `t` key without verbose `type`. */
    public static boolean isCompactWireForm(JSONObject json) {
        return json != null && json.has("t") && !json.has("type");
    }

    public static String extractMessageType(JSONObject json) {
        if (json == null) {
            return "";
        }
        if (json.has("type")) {
            return json.optString("type", "");
        }
        if (json.has("t")) {
            return json.optString("t", "");
        }
        return "";
    }

    public static JSONObject encode(JSONObject json) throws JSONException {
        if (json == null) {
            return null;
        }
        if (isCameraCommandJson(json.toString())) {
            return json;
        }
        String messageType = extractMessageType(json);
        if (!shouldCompactOutbound(messageType)) {
            return json;
        }
        return compactObject(json, true);
    }

    public static JSONObject encode(String jsonData) throws JSONException {
        if (jsonData == null || !jsonData.startsWith("{")) {
            return null;
        }
        return encode(new JSONObject(jsonData));
    }

    public static JSONObject decode(JSONObject json) throws JSONException {
        if (json == null) {
            return null;
        }
        if (isCameraCommandJson(json.toString())) {
            return json;
        }
        return expandObject(json, true);
    }

    /**
     * Expand compact wire JSON only when the message type is on the ROI allowlist.
     *
     * @return expanded JSON, passthrough expanded JSON, or {@code null} if compact form is unsupported
     */
    public static JSONObject decodeIfSupported(JSONObject json) throws JSONException {
        if (json == null) {
            return null;
        }
        if (isCameraCommandJson(json.toString())) {
            return json;
        }
        if (!isCompactWireForm(json)) {
            return json;
        }
        String compactType = json.optString("t", "");
        if (!supportsCompactInbound(compactType)) {
            return null;
        }
        return expandObject(json, true);
    }

    public static JSONObject decode(String jsonData) throws JSONException {
        if (jsonData == null || !jsonData.startsWith("{")) {
            return null;
        }
        return decode(new JSONObject(jsonData));
    }

    private static void putKey(String longKey, String shortKey) {
        LONG_TO_SHORT.put(longKey, shortKey);
        SHORT_TO_LONG.put(shortKey, longKey);
    }

    private static void putEnum(
            Map<String, Integer> toInt, Map<Integer, String> fromInt, String value, int code) {
        toInt.put(value, code);
        fromInt.put(code, value);
    }

    private static JSONObject compactObject(JSONObject input, boolean topLevel) throws JSONException {
        JSONObject output = new JSONObject();
        String messageType = input.optString("type", input.optString("t", ""));
        boolean skipResolvedConfig = false;

        if (topLevel && shouldDiffResolvedConfig(messageType) && input.has("resolvedConfig")) {
            JSONObject resolvedConfig = input.getJSONObject("resolvedConfig");
            String hash = hashConfig(resolvedConfig);
            resolvedConfigByHash.put(hash, cloneObject(resolvedConfig));
            if (resolvedConfigSent && hash.equals(currentSessionResolvedConfigHash)) {
                output.put(KEY_RESOLVED_CONFIG_HASH, hash);
                skipResolvedConfig = true;
            } else {
                currentSessionResolvedConfigHash = hash;
                resolvedConfigSent = true;
            }
        }

        Iterator<String> keys = input.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            if (skipResolvedConfig && "resolvedConfig".equals(key)) {
                continue;
            }
            Object value = input.get(key);
            String shortKey = LONG_TO_SHORT.getOrDefault(key, key);
            Object compacted = compactValue(key, value, messageType);
            if (shouldOmit(compacted)) {
                continue;
            }
            output.put(shortKey, compacted);
        }

        if (topLevel) {
            compactTimestamp(output);
        }
        return output;
    }

    private static boolean shouldDiffResolvedConfig(String messageType) {
        return "photo_status".equals(messageType) || "stream_status".equals(messageType);
    }

    private static Object compactValue(String longKey, Object value, String messageType)
            throws JSONException {
        if (value == JSONObject.NULL || value == null) {
            return null;
        }
        if (value instanceof JSONObject) {
            return compactObject((JSONObject) value, false);
        }
        if (value instanceof JSONArray) {
            return compactArray((JSONArray) value);
        }
        if (value instanceof Boolean || value instanceof Number) {
            return value;
        }
        String str = String.valueOf(value);
        if ("status".equals(longKey) && "photo_status".equals(messageType)) {
            Integer code = PHOTO_STATUS_TO_INT.get(str);
            if (code != null) {
                return code;
            }
        }
        if ("kind".equals(longKey)) {
            Integer code = KIND_TO_INT.get(str);
            if (code != null) {
                return code;
            }
        }
        if ("source".equals(longKey)) {
            Integer code = SOURCE_TO_INT.get(str);
            if (code != null) {
                return code;
            }
        }
        if ("aeStateName".equals(longKey)) {
            Integer code = AE_STATE_TO_INT.get(str);
            if (code != null) {
                return code;
            }
        }
        return str;
    }

    private static JSONArray compactArray(JSONArray array) throws JSONException {
        JSONArray out = new JSONArray();
        for (int i = 0; i < array.length(); i++) {
            Object value = array.get(i);
            if (value instanceof JSONObject) {
                out.put(compactObject((JSONObject) value, false));
            } else {
                out.put(value);
            }
        }
        return out;
    }

    private static boolean shouldOmit(Object value) {
        if (value == null || value == JSONObject.NULL) {
            return true;
        }
        if (value instanceof Boolean && !((Boolean) value)) {
            return true;
        }
        if (value instanceof JSONObject && ((JSONObject) value).length() == 0) {
            return true;
        }
        if (value instanceof JSONArray && ((JSONArray) value).length() == 0) {
            return true;
        }
        return false;
    }

    private static void compactTimestamp(JSONObject output) throws JSONException {
        String tsKey = output.has("timestamp") ? "timestamp" : (output.has("ts") ? "ts" : null);
        if (tsKey == null) {
            return;
        }
        long absolute = output.optLong(tsKey);
        if (absolute <= 0) {
            return;
        }
        if (sessionConnectEpochMs <= 0) {
            sessionConnectEpochMs = absolute;
        }
        output.put("ts", absolute - sessionConnectEpochMs);
        if (!"ts".equals(tsKey)) {
            output.remove(tsKey);
        }
    }

    private static JSONObject expandObject(JSONObject input, boolean topLevel) throws JSONException {
        JSONObject output = new JSONObject();
        String messageType = input.optString("type", input.optString("t", ""));

        Iterator<String> keys = input.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object value = input.get(key);
            if (topLevel && "ts".equals(key)) {
                output.put("ts", value);
                continue;
            }
            String longKey = SHORT_TO_LONG.getOrDefault(key, key);
            Object expanded = expandValue(longKey, value, messageType);
            if (expanded == null || expanded == JSONObject.NULL) {
                continue;
            }
            output.put(longKey, expanded);
        }

        if (topLevel) {
            expandTimestamp(output);
            expandResolvedConfigHash(output);
            cacheResolvedConfigIfPresent(output);
        }
        return output;
    }

    private static void cacheResolvedConfigIfPresent(JSONObject output) throws JSONException {
        if (!output.has("resolvedConfig")) {
            return;
        }
        JSONObject resolvedConfig = output.getJSONObject("resolvedConfig");
        String hash = hashConfig(resolvedConfig);
        resolvedConfigByHash.put(hash, cloneObject(resolvedConfig));
        currentSessionResolvedConfigHash = hash;
        resolvedConfigSent = true;
    }

    private static void expandResolvedConfigHash(JSONObject output) throws JSONException {
        if (output.has("resolvedConfig") || !output.has(KEY_RESOLVED_CONFIG_HASH)) {
            return;
        }
        String hash = output.optString(KEY_RESOLVED_CONFIG_HASH, "");
        JSONObject cached = resolvedConfigByHash.get(hash);
        if (cached != null) {
            output.put("resolvedConfig", cloneObject(cached));
        }
        output.remove(KEY_RESOLVED_CONFIG_HASH);
    }

    private static Object expandValue(String longKey, Object value, String messageType)
            throws JSONException {
        if (value == JSONObject.NULL || value == null) {
            return null;
        }
        if (value instanceof JSONObject) {
            return expandObject((JSONObject) value, false);
        }
        if (value instanceof JSONArray) {
            JSONArray out = new JSONArray();
            JSONArray array = (JSONArray) value;
            for (int i = 0; i < array.length(); i++) {
                Object item = array.get(i);
                if (item instanceof JSONObject) {
                    out.put(expandObject((JSONObject) item, false));
                } else {
                    out.put(item);
                }
            }
            return out;
        }
        if ("status".equals(longKey) && value instanceof Number && "photo_status".equals(messageType)) {
            String restored = PHOTO_STATUS_FROM_INT.get(((Number) value).intValue());
            return restored != null ? restored : value;
        }
        if ("kind".equals(longKey) && value instanceof Number) {
            String restored = KIND_FROM_INT.get(((Number) value).intValue());
            return restored != null ? restored : value;
        }
        if ("source".equals(longKey) && value instanceof Number) {
            String restored = SOURCE_FROM_INT.get(((Number) value).intValue());
            return restored != null ? restored : value;
        }
        if ("aeStateName".equals(longKey) && value instanceof Number) {
            String restored = AE_STATE_FROM_INT.get(((Number) value).intValue());
            return restored != null ? restored : value;
        }
        return value;
    }

    private static void expandTimestamp(JSONObject output) throws JSONException {
        if (!output.has("timestamp") && output.has("ts")) {
            long delta = output.optLong("ts");
            long absolute = sessionConnectEpochMs > 0 ? sessionConnectEpochMs + delta : delta;
            output.put("timestamp", absolute);
            output.remove("ts");
            if (sessionConnectEpochMs <= 0) {
                sessionConnectEpochMs = absolute - delta;
            }
        }
    }

    public static String hashConfig(JSONObject config) {
        try {
            return fnv1a32Hex(canonicalObject(config).toString());
        } catch (JSONException e) {
            return fnv1a32Hex(config.toString());
        }
    }

    private static JSONObject canonicalObject(JSONObject input) throws JSONException {
        List<String> keys = new ArrayList<>();
        Iterator<String> it = input.keys();
        while (it.hasNext()) {
            keys.add(it.next());
        }
        Collections.sort(keys);
        JSONObject out = new JSONObject();
        for (String key : keys) {
            Object value = input.get(key);
            if (value instanceof JSONObject) {
                out.put(key, canonicalObject((JSONObject) value));
            } else {
                out.put(key, value);
            }
        }
        return out;
    }

    private static JSONObject cloneObject(JSONObject input) throws JSONException {
        return new JSONObject(input.toString());
    }

    private static String fnv1a32Hex(String value) {
        byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
        int hash = 0x811c9dc5;
        for (byte b : bytes) {
            hash ^= (b & 0xff);
            hash *= 0x01000193;
        }
        return String.format("%08x", hash);
    }
}
